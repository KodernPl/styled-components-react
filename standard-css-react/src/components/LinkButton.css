@import '../styles/variables.css';

.link-button {
  display: inline-block;
  width: 20rem;
  padding: .5rem 0;
  margin: .5rem 1rem;
  border-width: .2rem;
  border-radius: .3rem;
  border-style: solid;
  border-image: initial;
  cursor: pointer;
  text-align: center;
  text-decoration: none;
  font-size: 1.2rem;
  text-transform: uppercase;

  &--default--light {
    color: var(--color-palevioletred);
    border-color: var(--color-white);
    background: var(--color-white);
  }

  &--default--dark {
    color: var(--color-black);
    border-color: var(--color-white);
    background: var(--color-white);
  }

  &--transparent--light {
    color: var(--color-white);

    &:hover {
      background: var(--color-alpha-grey);
    }
  }

  &--transparent--dark {
    color: var(--color-white);
    border-color: var(--color-white);

    &:hover {
      background: var(--color-grey);
    }
  }

  &--transparentDark--light {
    color: var(--color-white);
    border-color: var(--color-white);
    background: var(--gradient-light);
  }

  &--transparentDark--dark {
    color: var(--color-white);
    border-color: var(--color-white);
    background: var(--gradient-dark);
  }
}
