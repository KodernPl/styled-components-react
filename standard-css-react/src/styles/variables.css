:root {
  --color-white: white;
  --color-gradient1: #667db6;
  --color-gradient2: #0082c8;
  --color-gradient3: #0082c8;
  --color-gradient4: #667db6;
  --color-white: #fff;
  --color-black: #000;
  --color-dark-grey: #222;
  --color-darkest-grey: #111;
  --color-grey: grey;
  --color-palevioletred: palevioletred;
  --color-sc-pink: rgb(219, 112, 147);
  --color-sc-orange: rgb(218, 163, 87);
  --color-alpha-black: rgba(0, 0, 0, 0.2);
  --color-alpha-grey: rgba(20, 20, 20, 0.1);
  --gradient-light: linear-gradient(to top right, rgb(219, 112, 147), rgb(218, 163, 87));
  --gradient-dark: linear-gradient(to top right, #667db6, #0082c8, #0082c8);
}

@custom-media --desktop (min-width: 1024px);
