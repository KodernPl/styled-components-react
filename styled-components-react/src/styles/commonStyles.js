import styled from 'styled-components';

export const CenterElements = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
`;

export const CustomLink = styled.a`
  display: inline-block;
  width: 20rem;
  padding: .5rem 0;
  margin: .5rem 1rem;
  border-width: .2rem;
  border-radius: .3rem;
  border-style: solid;
  border-image: initial;
  cursor: pointer;
  text-align: center;
  text-decoration: none;
  font-size: 1.2rem;
`;
