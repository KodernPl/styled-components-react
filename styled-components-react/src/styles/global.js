import { createGlobalStyle } from 'styled-components';

export default createGlobalStyle`
  html {
    box-sizing: border-box;
    font-size: 62.5%;
    font-family: sans-serif;
  }

  body {
    margin: 0;
    padding: 0;
  }

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }
`;
