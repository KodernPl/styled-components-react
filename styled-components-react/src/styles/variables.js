export const COLORS = {
  gradient1: '#667db6',
  gradient2: '#0082c8',
  gradient3: '#0082c8',
  gradient4: '#667db6',

  white: '#fff',
  black: '#000',
  darkGrey: '#222',
  darkestGrey: '#111',

  grey: 'grey',
  palevioletred: 'palevioletred',

  scPink: 'rgb(219, 112, 147)',
  scOrange: 'rgb(218, 163, 87)',

  alphaBlack: 'rgba(0, 0, 0, 0.2)',
  alphaGrey: 'rgba(20, 20, 20, 0.1)',
};

export const THEMES = {
  light: 'light',
  dark: 'dark',
};

export const GRADIENTS = {
  light: `linear-gradient(to top right, ${COLORS.scPink}, ${COLORS.scOrange})`,
  dark: `linear-gradient(to top right, ${COLORS.gradient1}, ${COLORS.gradient2}, ${COLORS.gradient4})`,
};

export const BREAKPOINTS = {
  desktop: '(min-width: 1024px)',
};
