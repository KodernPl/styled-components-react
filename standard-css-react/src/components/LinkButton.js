import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import { withTheme } from '../utils/themeContext';
import { THEMES } from '../styles/variables';

import './LinkButton.css';

export const LINK_BUTTON_VARIANTS = {
  transparent: 'transparent',
  transparentDark: 'transparentDark',
  default: 'default',
};

const ns = 'link-button';
const displayName = 'LinkButton';

const propTypes = {
  variant: PropTypes.string,
  theme: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

const defaultProps = {
  variant: 'default',
};

const LinkButton = (props) => {
  const { theme, variant, children } = props;
  const classNames = cx(
    [`${ns}`],
    {
      [`${ns}--${variant}--light`]: theme === THEMES.light,
      [`${ns}--${variant}--dark`]: theme === THEMES.dark,
    },
  );

  return <a className={classNames} {...props}>{children}</a>;
};

LinkButton.displayName = displayName;
LinkButton.propTypes = propTypes;
LinkButton.defaultProps = defaultProps;

export default withTheme(LinkButton);
