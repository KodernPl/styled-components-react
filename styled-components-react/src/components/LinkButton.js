import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import styledTheming from 'styled-theming';

import { CustomLink } from '../styles/commonStyles';
import { COLORS, GRADIENTS, THEMES } from '../styles/variables';

export const LINK_BUTTON_VARIANTS = {
  transparent: 'transparent',
  transparentDark: 'transparentDark',
  default: 'default',
};

const themesProps = styledTheming.variants('mode', 'variant', {
  [LINK_BUTTON_VARIANTS.default]: {
    [THEMES.light]: css`
      color: ${COLORS.palevioletred};
      border-color: ${COLORS.white};
      background: ${COLORS.white};
    `,
    [THEMES.dark]: css`
      color: ${COLORS.black};
      background: ${COLORS.white};
      border-color: ${COLORS.white};
    `,
  },
  [LINK_BUTTON_VARIANTS.transparent]: {
    [THEMES.light]: css`
      color: ${COLORS.white};

      &:hover {
        background: ${COLORS.alphaGrey};
      }
    `,
    [THEMES.dark]: css`
      color: ${COLORS.white};
      border-color: ${COLORS.white};

      &:hover {
        background: ${COLORS.grey};
      }
    `,
  },
  [LINK_BUTTON_VARIANTS.transparentDark]: {
    [THEMES.light]: css`
      color: ${COLORS.white};
      border-color: ${COLORS.white};
      background: ${GRADIENTS.light};
    `,
    [THEMES.dark]: css`
      color: ${COLORS.white};
      border-color: ${COLORS.white};
      background: ${GRADIENTS.dark};
    `,
  },
});

const StyledLink = styled(CustomLink)`
  ${themesProps}
  text-transform: uppercase;
`;

const displayName = 'LinkButton';

const propTypes = {
  variant: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
};

const defaultProps = {
  variant: 'default',
};

const LinkButton = (props) => {
  const { children } = props;
  return <StyledLink {...props}>{children}</StyledLink>;
};

LinkButton.displayName = displayName;
LinkButton.propTypes = propTypes;
LinkButton.defaultProps = defaultProps;

export default LinkButton;
