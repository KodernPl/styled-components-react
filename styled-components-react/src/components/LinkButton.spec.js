import React from 'react';
import { ThemeProvider } from 'styled-components';
import renderer from 'react-test-renderer';

import { THEMES } from '../styles/variables';

import LinkButton, { LINK_BUTTON_VARIANTS } from './LinkButton';

const mockHREF = 'LINK';

const mockComponent = (theme, variant = LINK_BUTTON_VARIANTS.default) => (
  <ThemeProvider theme={{ mode: theme }}>
    <LinkButton
      href={mockHREF}
      variant={variant}
    >
      TEST
    </LinkButton>
  </ThemeProvider>
);
[
  THEMES.light,
  THEMES.dark,
].forEach((theme) => {
  Object.values(LINK_BUTTON_VARIANTS).forEach((variant) => {
    test(`renders correct link with ${theme} theme and ${variant} variant`, () => {
      expect(renderer.create(mockComponent(theme, variant)).toJSON()).toMatchSnapshot();
    });
  });
});
