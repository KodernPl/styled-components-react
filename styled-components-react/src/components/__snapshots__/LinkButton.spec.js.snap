// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correct link with dark theme and default variant 1`] = `
.c0 {
  display: inline-block;
  width: 20rem;
  padding: .5rem 0;
  margin: .5rem 1rem;
  border-width: .2rem;
  border-radius: .3rem;
  border-style: solid;
  border-image: initial;
  cursor: pointer;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.2rem;
  color: #000;
  background: #fff;
  border-color: #fff;
  text-transform: uppercase;
}

<a
  className="c0"
  href="LINK"
>
  TEST
</a>
`;

exports[`renders correct link with dark theme and transparent variant 1`] = `
.c0 {
  display: inline-block;
  width: 20rem;
  padding: .5rem 0;
  margin: .5rem 1rem;
  border-width: .2rem;
  border-radius: .3rem;
  border-style: solid;
  border-image: initial;
  cursor: pointer;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.2rem;
  color: #fff;
  border-color: #fff;
  text-transform: uppercase;
}

.c0:hover {
  background: grey;
}

<a
  className="c0"
  href="LINK"
>
  TEST
</a>
`;

exports[`renders correct link with dark theme and transparentDark variant 1`] = `
.c0 {
  display: inline-block;
  width: 20rem;
  padding: .5rem 0;
  margin: .5rem 1rem;
  border-width: .2rem;
  border-radius: .3rem;
  border-style: solid;
  border-image: initial;
  cursor: pointer;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.2rem;
  color: #fff;
  border-color: #fff;
  background: linear-gradient(to top right,#667db6,#0082c8,#667db6);
  text-transform: uppercase;
}

<a
  className="c0"
  href="LINK"
>
  TEST
</a>
`;

exports[`renders correct link with light theme and default variant 1`] = `
.c0 {
  display: inline-block;
  width: 20rem;
  padding: .5rem 0;
  margin: .5rem 1rem;
  border-width: .2rem;
  border-radius: .3rem;
  border-style: solid;
  border-image: initial;
  cursor: pointer;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.2rem;
  color: palevioletred;
  border-color: #fff;
  background: #fff;
  text-transform: uppercase;
}

<a
  className="c0"
  href="LINK"
>
  TEST
</a>
`;

exports[`renders correct link with light theme and transparent variant 1`] = `
.c0 {
  display: inline-block;
  width: 20rem;
  padding: .5rem 0;
  margin: .5rem 1rem;
  border-width: .2rem;
  border-radius: .3rem;
  border-style: solid;
  border-image: initial;
  cursor: pointer;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.2rem;
  color: #fff;
  text-transform: uppercase;
}

.c0:hover {
  background: rgba(20,20,20,0.1);
}

<a
  className="c0"
  href="LINK"
>
  TEST
</a>
`;

exports[`renders correct link with light theme and transparentDark variant 1`] = `
.c0 {
  display: inline-block;
  width: 20rem;
  padding: .5rem 0;
  margin: .5rem 1rem;
  border-width: .2rem;
  border-radius: .3rem;
  border-style: solid;
  border-image: initial;
  cursor: pointer;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1.2rem;
  color: #fff;
  border-color: #fff;
  background: linear-gradient(to top right,rgb(219,112,147),rgb(218,163,87));
  text-transform: uppercase;
}

<a
  className="c0"
  href="LINK"
>
  TEST
</a>
`;
